File: sec-blueprints/service-to-service/api-onbehalfof-springsecurity/src/main/java/no/nav/security/examples/springsecurity/client/DownstreamResourceClient.java
package no.nav.security.examples.springsecurity.client;

import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Component;
import org.springframework.web.client.RestTemplate;

@Component
public class DownstreamResourceClient {

    private final String url;
    private final RestTemplate restTemplate;

    public DownstreamResourceClient(@Value("${downstream.resource.url}") String url, RestTemplate restTemplate) {
        this.url = url;
        this.restTemplate = restTemplate;
    }

    public String tokeninfo() {
        return restTemplate.getForObject(url + "/tokeninfo", String.class);
    }

    public String ping() {
        return restTemplate.getForObject(url, String.class);
    }
}

File: sec-blueprints/service-to-service/api-onbehalfof-ktor/src/main/kotlin/no/nav/security/examples/DownstreamResourceClient.kt
package no.nav.security.examples

import com.fasterxml.jackson.databind.JsonNode
import com.github.michaelbull.result.Err
import com.github.michaelbull.result.Ok
import com.github.michaelbull.result.Result
import com.github.michaelbull.result.andThen
import io.ktor.client.HttpClient
import io.ktor.client.request.get
import io.ktor.client.request.header
import io.ktor.http.HttpHeaders
import org.slf4j.LoggerFactory

private val logger = LoggerFactory.getLogger(DownstreamResourceClient::class.java)

class DownstreamResourceClient(
    private val azureAdClient: AzureAdClient,
    private val httpClient: HttpClient = defaultHttpClient
) {
    suspend fun get(
        resource: Resource,
        accessToken: String
    ): Result<Resource, ThrowableErrorMessage> {
        val scopes = listOf("api://${resource.clientId}/.default")
        return azureAdClient
            .getOnBehalfOfAccessTokenForResource(scopes, accessToken)
            .andThen { oboAccessToken -> fetchFromDownstreamApi(resource, oboAccessToken) }
            .andThen { response -> Ok(resource.addResponse(response)) }
    }

    private suspend fun fetchFromDownstreamApi(
        resource: Resource,
        oboAccessToken: AccessToken
    ): Result<JsonNode, ThrowableErrorMessage> =
        runCatching {
            httpClient.get<JsonNode>(resource.url) {
                header(HttpHeaders.Authorization, "Bearer ${oboAccessToken.accessToken}")
            }
        }.fold(
            onSuccess = { result -> Ok(result) },
            onFailure = { error ->
                logger.error("received error from downstream api", error)
                Err(ThrowableErrorMessage(message = "Error response from '${resource.url}'", throwable = error))
            }
        )
}
